File: src/ops/Pump.cpp
#include "chemsim/ops/Pump.hpp"
#include <stdexcept>
#include <cmath>

namespace chemsim {

Pump::Pump(const FlashCalculator& fc, double P_out, double eta)
    : fc_(fc), P_out_(P_out), eta_(eta) {
    if (eta <= 0.0 || eta > 1.0)
        throw std::invalid_argument("Pump: eta must be in (0, 1]");
}

void Pump::solve() {
    if (inlet.totalFlow <= 0.0)
        throw std::invalid_argument("Pump: inlet.totalFlow must be > 0");
    if (P_out_ <= inlet.P)
        throw std::invalid_argument("Pump: P_out must be greater than inlet pressure");

    const double T_in = inlet.T;
    const double P_in = inlet.P;
    const auto&  z    = inlet.z;

    // Inlet enthalpy (liquid assumed for pump)
    double H_in = fc_.phaseEnthalpy(T_in, P_in, z, true);
    double S_in = fc_.phaseEntropy (T_in, P_in, z, true);

    // Isentropic outlet: flashPS at P_out with S_in
    auto r_s = fc_.flashPS(P_out_, S_in, z, T_in);
    double H_out_s = fc_.totalEnthalpy(r_s);

    // Actual enthalpy: W_actual = W_isentropic / eta
    double W_s   = H_out_s - H_in;          // J/mol, isentropic work
    double W_act = W_s / eta_;               // actual work per mol
    double H_out = H_in + W_act;

    // Actual outlet: flashPH at P_out with H_out
    auto r_out = fc_.flashPH(P_out_, H_out, z, r_s.T);

    outlet           = inlet;                // copy name, z, totalFlow
    outlet.P         = r_out.P;
    outlet.T         = r_out.T;
    outlet.vaporFraction = r_out.beta;
    outlet.x         = r_out.x;
    outlet.y         = r_out.y;
    outlet.phase     = (r_out.beta < 1e-10)  ? Phase::LIQUID
                     : (r_out.beta > 1-1e-10) ? Phase::VAPOR
                                               : Phase::MIXED;
    outlet.H = fc_.totalEnthalpy(r_out);
    outlet.S = fc_.totalEntropy (r_out);

    shaft_work_mol  = W_act;
    shaft_power_W   = inlet.totalFlow * W_act;
}

} // namespace chemsim

File: chemsim/ops/pump.py
"""Isentropic pump with efficiency."""
from __future__ import annotations

from chemsim.core import Phase, Stream
from chemsim.ops.base import IUnitOp
from chemsim.thermo.flash import FlashCalculator


class PumpOp(IUnitOp):
    def __init__(self, fc: FlashCalculator, P_out: float, eta: float = 0.75):
        if not (0.0 < eta <= 1.0):
            raise ValueError("PumpOp: eta must be in (0, 1]")
        self._fc = fc
        self._P_out = P_out
        self._eta = eta
        self.inlet = Stream()
        self.outlet = Stream()
        self.shaft_work_mol: float = 0.0
        self.shaft_power_W: float = 0.0

    def inlet_ports(self):
        return ["in"]

    def outlet_ports(self):
        return ["out"]

    def set_inlet(self, port: str, stream: Stream) -> None:
        if port != "in":
            raise ValueError(f"PumpOp: unknown inlet port '{port}'")
        self.inlet = stream

    def get_outlet(self, port: str) -> Stream:
        if port != "out":
            raise ValueError(f"PumpOp: unknown outlet port '{port}'")
        return self.outlet

    def solve(self) -> None:
        if self.inlet.total_flow <= 0.0:
            raise ValueError("PumpOp: inlet.total_flow must be > 0")
        if self._P_out <= self.inlet.P:
            raise ValueError("PumpOp: P_out must be greater than inlet pressure")

        T_in = self.inlet.T; P_in = self.inlet.P; z = self.inlet.z
        H_in = self._fc.phase_enthalpy(T_in, P_in, z, True)
        S_in = self._fc.phase_entropy(T_in, P_in, z, True)

        r_s = self._fc.flash_PS(self._P_out, S_in, z, T_in)
        H_out_s = self._fc.total_enthalpy(r_s)

        W_s = H_out_s - H_in
        W_act = W_s / self._eta
        H_out = H_in + W_act

        r_out = self._fc.flash_PH(self._P_out, H_out, z, r_s.T)

        import copy
        self.outlet = copy.copy(self.inlet)
        self.outlet.P = r_out.P; self.outlet.T = r_out.T
        self.outlet.vapor_fraction = r_out.beta
        self.outlet.x = r_out.x; self.outlet.y = r_out.y
        self.outlet.phase = (Phase.LIQUID if r_out.beta < 1e-10
                              else Phase.VAPOR if r_out.beta > 1.0 - 1e-10
                              else Phase.MIXED)
        self.outlet.H = self._fc.total_enthalpy(r_out)
        self.outlet.S = self._fc.total_entropy(r_out)
        self.shaft_work_mol = W_act
        self.shaft_power_W  = self.inlet.total_flow * W_act
